### ISSUE TITLE
Random seed resets when loading checkpoint

### ISSUE THREAD

**Issue by keyonvafa** [tensorflow/probability#1286]
I'm noticing that restoring a loaded checkpoint leads to undesirable behavior if we have set a global random seed. Namely, the random seed state resets, so the same random samples from earlier in training are re-used. This artificially deflates the loss, especially for some Bayesian inference methods. Is there any way to store and load the current random state with the checkpoint? (This would also be helpful for reproducibility.)

For example, in the code I've pasted below, I'm minimizing KL(q(x) || p(x)) with respect to mu and sigma, where p(x) ~ LogNormal(0.3, 0.3) and q(x) ~ Lognormal(mu, sigma). The code works fine, until we stop running and restore from a checkpoint. When we restore, the loss is artificially deflated, because the same random samples from earlier in training are being used. This only happens when we set a global random seed. If I remove the `tf.random.set_seed` line below (or change the seed), everything works as expected.

I've attached a Tensorboard picture of the loss that illustrates what's going on. I separately train a run that includes the global random seed, and a run that doesn't include the global random seed. For both runs, I stop training about halfway through and restart at the checkpoint. The run with the random seed (light blue) takes a large spike once we restore; this spike does not appear for the run without the random seed (pink). 

<img width="388" alt="Screen Shot 2021-03-29 at 2 49 59 PM" src="https://user-images.githubusercontent.com/10766891/112884894-0fd85380-909e-11eb-880d-fe27c8a83f51.png">

Here's the code. This is all using Python 3.9.2, Tensorflow 2.4.1, and Tensorflow Probability 0.12.1 on a single NVIDIA Titan V GPU with CUDA 10.2.

Thank you!

``` python
class Net(tf.keras.Model):
  def __init__(self):
    super(Net, self).__init__()
    self.dim = 1000000
    self.location = tf.Variable(
      tf.keras.initializers.GlorotUniform()(shape=[self.dim]))
    self.scale = tfp.util.TransformedVariable(
      tf.ones([self.dim]),
      bijector=tfp.bijectors.Softplus())
    self.variational_dist = tfp.distributions.LogNormal(loc=self.location,
                                                        scale=self.scale)
  
  def call(self, input=None):
    prior_dist = tfp.distributions.Gamma(concentration=0.3, rate=0.3)
    # Compute ELBO by taking samples and reparameterization gradients.
    variational_samples = self.variational_dist.sample()
    log_prior = tf.reduce_sum(prior_dist.log_prob(variational_samples))
    entropy = -tf.reduce_sum(self.variational_dist.log_prob(variational_samples))
    elbo = log_prior + entropy
    return elbo, log_prior, entropy


def train_step(model, optim):
  with tf.GradientTape() as tape:
    elbo, log_prior, entropy = model(None)
    loss = -elbo
  trainable_variables = tape.watched_variables()
  grads = tape.gradient(loss, trainable_variables)
  optim.apply_gradients(zip(grads, trainable_variables))
  return loss, log_prior, entropy


def main(argv):
  del argv
  tf.random.set_seed(123)
  net = Net()
  net.start_epoch = tf.Variable(-1)
  optim = tf.optimizers.Adam(learning_rate=FLAGS.learning_rate)

  checkpoint = tf.train.Checkpoint(step=tf.Variable(-1), optimizer=optim, net=net)
  manager = tf.train.CheckpointManager(checkpoint, "./ckpts", max_to_keep=1)
  checkpoint.restore(manager.latest_checkpoint)

  for epoch in range(net.start_epoch.numpy() + 1, 1000):
    checkpoint.step.assign(epoch)
    loss, log_prior, entropy = train_step(net, optim)
    print("Epoch: {:>3d} Loss: {:.3f}".format(epoch, loss.numpy()))

    net.start_epoch.assign(epoch)
    manager.save()


if __name__ == "__main__":
  app.run(main)
```

**Comment by brianwa84** [comment]
TF stateful randomness holds state in RAM, so when you restart the process you restart the RNGs from the same place as your earlier run *started*. https://github.com/tensorflow/probability/blob/master/PRNGS.md

A few ideas come to mind. Probably simplest, you could move the `tf.random.set_seed` to the top of your loop: `tf.random.set_seed(epoch)`. Alternatively, since TFP now supports stateless samplers, you might add a seed argument to train_step, and write `step_seed, seed = tfp.random.split_seed(seed)` for each iteration.

**Comment by brianwa84** [comment]
Also, when you don't specify a seed, I guess TF may pull from /dev/urandom or elsewhere for better randomness.
Interesting that the loss is better when you are repeating samples. I guess that makes some sense, since the weights have already seen gradients from those samples.

**Comment by keyonvafa** [comment]
Great, thank you! That's super helpful. Moving `tf.random.set_seed` to the top of the loop appears to only have reproducible randomness when there's no `@tf.function` decorator, so I'll go with the second approach. Do you know what's the best way to save the seed as part of the checkpoint? When I try to save it via

``` python
_, seed = tfp.random.split_seed(0)
tf.train.Checkpoint(optimizer=optim, net=net, seed=seed)
```

I get `Checkpoint was expecting a trackable object`. When I try to save it as an attribute of the model (i.e. `net.seed = seed`), it isn't restored when loading the checkpoint.

**Comment by brianwa84** [comment]
I guess you'd have to assign it to a Variable like you did with epoch
number.

**Comment by keyonvafa** [comment]
Thank you!
